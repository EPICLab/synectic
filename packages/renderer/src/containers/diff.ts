import {diffLines} from 'diff';

/**
 * Diffs two blocks of text, comparing line by line, based on the algorithm proposed
 * in "An O(ND) Difference Algorithm and its Variations" (Myers, 1986). Results are formatted
 * according to the unified diff patch used in the Git protocol, where +/- indicate lines
 * added or removed in order to transform from original to updated.
 * @param original The base version of code or text to evaluate.
 * @param updated The updated version of code or text to compare against original.
 * @returns {string} A diff patch that conforms to the unified diff patch format.
 */
export const diff = (original: string, updated: string): string => {
  const diffLineChanges = diffLines(original, updated);
  let readableDiff = '';

  diffLineChanges.map(segment => {
    const segmentLines = segment.value.split(/\n/g, segment.count);
    if (segment.added) {
      segmentLines.map(line => (readableDiff += `+${line}\n`));
    } else if (segment.removed) {
      segmentLines.map(line => (readableDiff += `-${line}\n`));
    } else {
      segmentLines.map(line => (readableDiff += ` ${line}\n`));
    }
  });

  return readableDiff;
};
