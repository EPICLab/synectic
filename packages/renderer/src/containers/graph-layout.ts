import type {UUID} from '@syn-types/app';
import dagre from 'dagre';
import type {Edge, Node} from 'reactflow';
import {MarkerType, isEdge, isNode} from 'reactflow';
import type {CommitGraph, CommitVertex} from './hooks/useGitGraph';

const nodesep = 80;

const layoutGraph = (graph: CommitGraph, topological: UUID[]): (Edge | Node)[] => {
  const elements: (Edge | Node)[] = [];
  for (const oid of topological) {
    const commit = graph.get(oid.toString());
    if (commit) {
      const node: Node = getNode(commit);
      const edges: Edge[] = getEdges(commit);
      elements.push(node);
      edges.map(edge => elements.push(edge));
    }
  }
  return layoutOptimizer(elements);
};

const layoutOptimizer = (rfGraph: (Edge | Node)[]): (Edge | Node)[] => {
  const graph = new dagre.graphlib.Graph();
  graph.setGraph({nodesep: nodesep});
  graph.setDefaultEdgeLabel(() => {
    return {};
  });

  rfGraph.filter(isNode).map(node => graph.setNode(node.id, {width: 5, height: 5}));
  rfGraph.filter(isEdge).map(edge => graph.setEdge(edge.source, edge.target));
  dagre.layout(graph);

  return rfGraph.map(els => {
    if (isNode(els)) {
      const graphNode = graph.node(els.id);
      return {...els, position: {x: graphNode.x, y: graphNode.y}};
    }
    return els;
  });
};

const getNode = (commit: CommitVertex): Node<CommitVertex> => {
  return {
    id: commit.oid.toString(),
    type: 'gitNode',
    data: commit,
    position: {x: 0, y: 0},
  };
};

const getEdges = (commit: CommitVertex): Edge[] => {
  return commit.parents
    .filter(parent => parent.length > 0)
    .map(parent => ({
      id: `e${parent.toString().slice(0, 7)}=${
        commit.staged ? commit.oid.toString() : commit.oid.toString().slice(0, 7)
      }`,
      source: parent.toString(),
      target: commit.oid.toString(),
      animated: commit.staged ? true : false,
      arrowHeadType: MarkerType.ArrowClosed,
    }));
};

export default layoutGraph;
