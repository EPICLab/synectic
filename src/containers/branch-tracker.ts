import { Action } from 'redux';
import { ThunkAction } from 'redux-thunk';

import { RootState } from '../store/root';
import { loadCard } from './handlers';
import { getMetafile } from './metafiles';

export const loadBranchVersions = (): ThunkAction<Promise<void>, RootState, undefined, Action> => {
  return async (dispatch) => {
    const metafile = await dispatch(getMetafile({ virtual: { name: 'Version Tracker', handler: 'Tracker' } }));
    if (metafile) dispatch(loadCard({ metafile: metafile }));
  };
}