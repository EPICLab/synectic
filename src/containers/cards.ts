import { DateTime } from 'luxon';
import { v4 } from 'uuid';

import type { Card } from '../types';
import { ActionKeys, NarrowActionType } from '../store/actions';
import { HandlerRequiredMetafile } from './handlers';

type AddCardAction = NarrowActionType<ActionKeys.ADD_CARD>;
type UpdateCardAction = NarrowActionType<ActionKeys.UPDATE_CARD>;

/**
 * Action Creator for composing a valid ADD_CARD Redux Action.
 * @param metafile The related `Metafile` object containing a valid `handler` field.
 * @return An `AddCardAction` object that can be dispatched via Redux.
 */
export const addCard = (metafile: HandlerRequiredMetafile): AddCardAction => {
  const card: Card = {
    id: v4(),
    name: metafile.name,
    created: DateTime.local(),
    modified: metafile.modified,
    left: 10,
    top: 25,
    type: metafile.handler,
    metafile: metafile.id
  };
  return {
    type: ActionKeys.ADD_CARD,
    id: card.id,
    card: card
  };
}

/**
 * Action Creator for composing a valid UPDATE_CARD Redux Action. If the current Redux store does not contain a
 * matching card (based on UUID) for the passed parameter, then dispatching this action will not result in any
 * changes in the Redux store state.
 * @param card A `Card` object containing new field values to be updated.
 * @return An `UpdateCardAction` object that can be dispatched via Redux, including an updated timestamp in the
 * `modified` field.
 */
export const updateCard = (card: Card): UpdateCardAction => {
  return {
    type: ActionKeys.UPDATE_CARD,
    id: card.id,
    card: { ...card, modified: DateTime.local() }
  }
}