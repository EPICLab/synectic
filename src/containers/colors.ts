export type ColorSet = {
  primary: string;
  secondary: string;
};

// colors from http://clrs.cc/
export const colorSets: { [color: string]: ColorSet } = {
  navy: { primary: '#001f3f', secondary: '#80b5ff' },
  blue: { primary: '#0074d9', secondary: '#b3dbff' },
  aqua: { primary: '#7fdbff', secondary: '#004966' },
  teal: { primary: '#39cccc', secondary: '#000000' },
  olive: { primary: '#3d9970', secondary: '#163728' },
  green: { primary: '#2ecc40', secondary: '#0e3e14' },
  lime: { primary: '#01ff70', secondary: '#00662c' },
  yellow: { primary: '#ffdc00', secondary: '#665800' },
  orange: { primary: '#ff851b', secondary: '#663000' },
  red: { primary: '#ff4136', secondary: '#800600' },
  maroon: { primary: '#85144b', secondary: '#eb7ab1' },
  fuchsia: { primary: '#f012be', secondary: '#65064f' },
  purple: { primary: '#b10dc9', secondary: '#efa9f9' },
  black: { primary: '#111111', secondary: '#dddddd' },
  gray: { primary: '#aaaaaa', secondary: '#000000' },
  silver: { primary: '#dddddd', secondary: '#000000' },
  white: { primary: '#ffffff', secondary: '#444444' },
  conflictRed: { primary: '#da6473', secondary: '' },
  stagedBlue: { primary: '#61aeee', secondary: '' },
  unstagedOrange: { primary: '#d19a66', secondary: '' },
  nominalGreen: { primary: '#50c878', secondary: '' }
};

export const getColor = (color: keyof typeof colorSets): ColorSet => {
  return colorSets[color];
};
