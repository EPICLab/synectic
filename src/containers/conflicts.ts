import * as path from 'path';
import { PathLike } from 'fs-extra';
import * as io from './io';
import { getIgnore } from './git-plumbing';
import { asyncFilter, removeUndefined } from './format';
import { Ignore } from 'ignore';
import { createAsyncThunk } from '@reduxjs/toolkit';
import { AppThunkAPI } from '../store/hooks';
import { v4 } from 'uuid';
import { fetchMetafile } from '../store/thunks/metafiles';
import { DateTime } from 'luxon';
import { loadCard } from '../store/thunks/handlers';
import { getBranchRoot, getRoot, getWorktreePaths } from './git-path';

export type Conflict = {
    /** The relative or absolute path to the file containing conflicts. */
    filepath: PathLike,
    /** An array of indexed conflicts indicating the starting position of each conflict in the file. */
    conflicts: number[]
}

export const loadConflictManagers = createAsyncThunk<void, void, AppThunkAPI>(
    'conflicts/loadConflictManagers',
    async (_, thunkAPI) => {
        const repos = removeUndefined(Object.values(thunkAPI.getState().repos.entities));
        await Promise.all(repos.map(async repo => {
            await Promise.all(repo.local.map(async branchId => {
                const branch = thunkAPI.getState().branches.entities[branchId];
                const root = branch ? await getBranchRoot(repo.root, branch.ref) : undefined;
                const conflicts = await checkProject(root);
                if (branch && root && conflicts.length > 0) {
                    const { base, compare } = await resolveConflictBranches(root);
                    const conflictManager = await thunkAPI.dispatch(fetchMetafile({
                        virtual: {
                            id: v4(),
                            modified: DateTime.local().valueOf(),
                            name: `Conflicts`,
                            handler: 'ConflictManager',
                            repo: repo.id,
                            path: root,
                            merging: { base: (base ? base : branch.ref), compare: compare }
                        }
                    })).unwrap();
                    await thunkAPI.dispatch(loadCard({ metafile: conflictManager }));
                }
            }));
        }));
    }
);

export const resolveConflictBranches = async (root: PathLike): Promise<{ base: string | undefined, compare: string }> => {
    const branchPattern = /(?<=^Merge branch(es)? .*)('.+?')+/gm;
    const worktree = await getWorktreePaths(root);
    const mergeMsg = worktree.gitdir ? await io.readFileAsync(path.join(worktree.gitdir.toString(), 'MERGE_MSG'), { encoding: 'utf-8' }) : '';
    const match = mergeMsg.match(branchPattern);
    return match
        ? match.length === 2
            ? { base: match[0].replace(/['"]+/g, ''), compare: match[1].replace(/['"]+/g, '') }
            : { base: undefined, compare: match[0].replace(/['"]+/g, '') }
        : { base: undefined, compare: '' };
}

export const checkFilepath = async (filepath: PathLike, ignoreManager?: Ignore): Promise<Conflict | undefined> => {
    const root = await getRoot(filepath);
    if (root && !(await io.isDirectory(filepath))) {
        const conflictPattern = /<<<<<<<[^]+?=======[^]+?>>>>>>>/gm;
        const ignore = ignoreManager ? ignoreManager : (await getIgnore(root, true));
        if (ignore.ignores(path.relative(root.toString(), filepath.toString()))) return undefined;
        const content = await io.readFileAsync(filepath, { encoding: 'utf-8' });
        const matches = Array.from(content.matchAll(conflictPattern));
        const matching = { filepath: filepath, conflicts: removeUndefined(matches.map(m => m.index)) };
        if (matching.conflicts.length > 0) return matching;
    }
    return undefined;
}

export const checkProject = async (root: PathLike | undefined): Promise<Conflict[]> => {
    if (root) {
        const conflictPattern = /<<<<<<<[^]+?=======[^]+?>>>>>>>/gm;
        const ignore = (await getIgnore(root, true));

        const paths = (await io.readDirAsyncDepth(root, 1))
            .filter(p => p !== root)                                          // filter root filepath from results
            .filter(p => !ignore.ignores(path.relative(root.toString(), p))); // filter based on git-ignore rules
        const dirpaths = await asyncFilter(paths, async p => io.isDirectory(p));
        const filepaths = paths.filter(p => !dirpaths.includes(p));

        const matching = await Promise.all(filepaths.map(async f => {
            const content = await io.readFileAsync(f, { encoding: 'utf-8' });
            const matches = Array.from(content.matchAll(conflictPattern));
            return { filepath: f, conflicts: removeUndefined(matches.map(m => m.index)) };
        }));

        return matching.filter(match => match.conflicts.length > 0);
    } else {
        return [];
    }
}