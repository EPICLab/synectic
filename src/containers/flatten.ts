export type ArrayElementType<T> = T extends Array<infer U> ? ArrayElementType<U> : T;
export type ObjectElementType<T> = T extends Record<string | number | symbol, infer U>
  ? ObjectElementType<U>
  : T;

/**
 * Flattens an n-depth array into a single level array containing all sub-array elements. Solutions
 * for the problem of flattening n-depth arrays in JavaScript are fairly simple, but the TypeScript
 * compiler requires closed type guarding in order to bubble up sub-types from within the sub-array
 * elements. Luckily, TypeScript 4.1 includes recursive conditional types (
 * [TS4.1 Recursive Conditional Types](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#recursive-conditional-types)),
 * which allows for deferred type checking while recursive sub-typing resolves the type
 * intersections for individual element depths. Some type casting is required (but should be
 * type-safe) in order to handle the gradual typing of `<unknown>` within the reducer function, but
 * this is an issue that could be solved in the future if TypeScript adds gradual inferred typing.
 *
 * Originally, this solution was a partial type enhancement that resulted in a `<any[]>`-typed
 * return array. This previous solution was type-aware and used an efficient recursive reduce
 * function in order to prevent non-array inputs, but could only guarantee that the output is an
 * array (prior to TS 4.1, we could not recursively infer the types of each sub-array and bubble
 * those types up to a combined intersection type). This solution was originally posted by
 * [`@maroun-baydoun`](@link https://qiita.com/maroun-baydoun/items/dbd65b56b3ef2365f583).
 * @param arr Array containing primitive values, objects, and other arrays.
 * @returns {ArrayElementType[]} Array of depth 1, containing all sub-array elements.
 */
export const flattenArray = <T extends unknown[]>(arr: T): ArrayElementType<T>[] => {
  const reducer = (prev: ArrayElementType<T>[], curr: unknown): ArrayElementType<T>[] => {
    return Array.isArray(curr)
      ? (curr as ArrayElementType<T>[]).reduce(reducer, prev)
      : [...prev, curr as ArrayElementType<T>];
  };
  return arr.reduce<ArrayElementType<T>[]>(reducer, []);
};

/**
 * Flattens an n-depth object into a single level object containing all sub-object elements. When
 * flattening key-value pairs that contain an object value, it is necessary to discard the key name
 * of the outer container object and only hoist the inner pair. Some type casting is required (but
 * should be type-safe) in order to handle the gradual typing of results from square bracket
 * property accessors.
 * @param obj Object containing key-value pairs that point to primitives, arrays, and other objects.
 * @returns {Record<string | number | symbol, ObjectElementType>} d Object of depth 1, containing
 * all sub-object elements (excluding keys for container objects).
 */
export const flattenObject = <T extends Record<string | number | symbol, unknown>>(
  obj: T
): Record<string | number | symbol, ObjectElementType<T>> => {
  const flattened: Record<string, ObjectElementType<T>> = {};
  Object.keys(obj).forEach(key => {
    if (typeof obj[key] === 'object' && obj[key] !== null) {
      Object.assign(flattened, flattenObject(obj[key] as Record<string, ObjectElementType<T>>));
    } else {
      flattened[key] = obj[key] as ObjectElementType<T>;
    }
  });
  return flattened;
};
