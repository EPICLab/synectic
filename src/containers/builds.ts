import { exec } from 'child_process';
import util from 'util';
import { checkout } from 'isomorphic-git';
import * as fs from 'fs-extra';
import { join } from 'path';

import type { Repository } from '../types';
import { clone, merge } from '../containers/git-porcelain';
import { readDirAsync } from './io';
import { removeUndefinedProperties } from './format';

const promiseExec = util.promisify(exec);

export const tempClone = async (repo: Repository, branch: string): Promise<string> => {
  const cloneRoot = join(repo.root.toString(), '../', '.syn');
  await clone({ repo: repo, dir: cloneRoot, ref: branch });
  return cloneRoot;
}

export const build = async (repo: Repository, base: string, compare: string): Promise<{ installCode: number; buildCode: number; }> => {
  const cloneRoot = await tempClone(repo, base);
  await checkout({ fs: fs, dir: cloneRoot, ref: base });

  const mergeResult = await merge(cloneRoot, base, compare);
  const ref = removeUndefinedProperties({ ref: mergeResult.oid });
  // this next step seems unnecessary, but isomorphic-git.merge stages a reversal changeset that negates the results of the merge
  // reverting the target branch back to the merged commit (via checkout) allows these reversal changesets to be removed
  await checkout({ fs: fs, dir: cloneRoot, ...ref });

  const rootFiles = await readDirAsync(cloneRoot);
  const packageManager = rootFiles.find(file => file === 'yarn.lock') ? 'yarn' : 'npm';

  let [installCode, buildCode] = [-1, -1];
  try {
    const installResults = promiseExec(`${packageManager} install`, { cwd: cloneRoot });
    installResults.child.stdout?.on('data', data => console.log('INSTALL: ' + data));
    installResults.child.stderr?.on('data', data => console.log('INSTALL error: ' + data));
    installResults.child.on('close', code => {
      console.log(`INSTALL 'close' listener found code: ${code}`);
      (code ? (installCode = code) : null);
    });
    installResults.child.on('exit', code => {
      console.log(`INSTALL 'exit' listener found code: ${code}`);
      (code ? (installCode = code) : null);
    });
    installResults.child.on('error', error => {
      console.log('INSTALL \'error\' listener found error:');
      console.log({ error });
    });
    await installResults;
  } catch (e) {
    console.log('INSTALL ERROR');
    console.log(e);
  }

  if (installCode === 0) {
    const packageManagerBuildScript = packageManager === 'yarn' ? 'run' : 'run-script';
    try {
      const buildResults = promiseExec(`${packageManager} ${packageManagerBuildScript} build`, { cwd: cloneRoot });
      buildResults.child.stdout?.on('data', data => console.log('BUILD: ' + data));
      buildResults.child.stderr?.on('data', data => console.log('BUILD error: ' + data));
      buildResults.child.on('close', code => {
        console.log(`BUILD 'close' listener found code: ${code}`);
        (code ? (buildCode = code) : null);
      });
      buildResults.child.on('exit', code => {
        console.log(`BUILD 'exit' listener found code: ${code}`);
        (code ? (buildCode = code) : null);
      });
      buildResults.child.on('error', error => {
        console.log('BUILD \'error\' listener found error:');
        console.log({ error });
      });
      await buildResults;
    } catch (e) {
      console.log('BUILD ERROR');
      console.error(e);
    }
  }

  fs.remove(cloneRoot, (error) => console.log(error));

  return { installCode: installCode, buildCode: buildCode };
}