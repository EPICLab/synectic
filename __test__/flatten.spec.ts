import { flattenArray, flattenObject } from '../src/containers/flatten';

describe('flatten.flatten', () => {

  it('flatten resolves simple arrays of depth 1', () => {
    expect(flattenArray([1])).toStrictEqual([1]);
  });

  it('flatten resolves arrays of equal max-depth', () => {
    expect(flattenArray([[1], [2]])).toStrictEqual([1, 2]);
  });

  it('flatten resolves arrays of non-equal max-depth', () => {
    expect(flattenArray([[1], 2])).toStrictEqual([1, 2]);
  });

  it('flatten resolves heterogenous arrays of primitive value types', () => {
    expect(flattenArray([[1, 2], [['apple'], [true]]])).toStrictEqual([1, 2, 'apple', true]);
  });

  it('flatten resolves heterogenous arrays of object types', () => {
    const flatArray = [{ id: 1 }, { 2: 'pink' }, 'apple', { state: true }];
    expect(flattenArray([[{ id: 1 }, { 2: 'pink' }], [['apple'], [{ state: true }]]])).toStrictEqual(flatArray);
  });

});

describe('flatten.flattenObject', () => {
  it('flattenObject resolves nested heterogenous JavaScript objects', () => {
    const origObj = { a: 'string', b: { c: 13 } };
    expect(flattenObject(origObj)).toMatchSnapshot();
  });
});