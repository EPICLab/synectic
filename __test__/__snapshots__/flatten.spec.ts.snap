// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flatten.flattenObject flattenObject resolves nested heterogenous JavaScript objects 1`] = `
Object {
  "a": "string",
  "c": 13,
}
`;
